### DISCUSSION TITLE
DATAVIEW alternatives in nvim

### DISCUSSION THREAD

**Discussion by Dardanos-Aeolus** [epwalsh/obsidian.nvim#713] [Cookbook]
Hi everybody,

Lots of people use DATAVIEW as a plugin in Obsidian.
I haven't found some similar for Neovim.

So I wanted to start a discussion and give a starting point for people, who are inexperienced like me :-)
(I know this is somewhat hacky - please improve and comment if you can )

### 1. A simple customised telescope prompt
A telescope prompt could be an alternative for very simple Dataview usage.

So for example I want to have a list of all my notes with active projects.
These have a frontmatter with:
``` markdown
---
type: project
status: open
---
```
in init.lua:
``` markdown

projekt_liste_picker = function()
	require("telescope.builtin").live_grep({
	cwd="/path/to/your/vault", 
        -- this is the regex search with two keywords
	default_text="type: project.*status: open",
	-- I like normal mode, so I can go up/down with j/k quickly
        initial_mode = "normal",
	vimgrep_arguments={
		"rg",
        	"--color=never",
		"-l",     -- list matching file only fonce
                -- I don't want templates to show, so I exclude these here:
		"--iglob=!templates",
          	"--no-heading",
		"--multiline",
		"--multiline-dotall",
          	"--with-filename",
          	"--line-number",
         	"--column",
          	"--smart-case",
        }
	})
```

### 2. A Python Script
Now this is much better :-D
Same application: I want to have a list with active projects.
Here it is written into a md-file inside my vault adding additional metadata.

Setup of python:
``` bash
pip install python-frontmatter
pip install PTable
```

``` python
import os
import frontmatter
from prettytable import PrettyTable, ALL, HEADER
outputfile_path = "/path/to/your/vault/Projektlist_dynamic.md"

table = PrettyTable()
table.field_names = ["Projectnumber", "Customer", "Status", "Link"]
table.hrules = HEADER
table.junction_char = '|'
table.align = 'l'

folder = "/path/to/your/vault"

for filename in os.listdir(folder):
    if 'md' in filename:
       filename_path = os.path.join(folder,filename)
       tmp_fm = frontmatter.load(filename_path, encoding='utf-8')
       
       if tmp_fm.get('type', 'No type metadatafield') == 'projekt':
           p_status = tmp_fm.get('status', 'ERROR55')
           if p_status == 'open' or p_status == 'hold':
               table.add_row([tmp_fm.get('project', '/'), tmp_fm.get('customer', 'N/A'), tmp_fm.get('status', '/'), r'[[' + filename + r']]'])


output = table.get_string()
with open(outputfile_path, "w", encoding='utf-8') as f:
    f.write("# Projectlist\n")
    f.write("*This File is genereated automatically -- all edits will be lost*\n\n")
    f.write(output)
```
OUTPUT:
``` markdown
# Projectlist
*This File is genereated automatically -- all edits will be lost*

| Projectnumber  | Customer        | Status | Link                                     |
|----------------|-----------------|--------|------------------------------------------|
| PR4500090      | Random customer | open   | [[20230921-15-13 Random_cu.md]]          |
| PR4500095 / 96 | Customer 1      | open   | [[20240816-15-57-customer1.md]]          |
| digital_Market | Internal        | hold   | [[20240830-10-23-landing-dig-market.md]] |
| PR4500120      | Forever Young   | open   | [[20240906-16-05-forever_young.md]]      |
| Must have      | Personal        | open   | [[20240911-16-09-must_have_urgent.md]]   |
```

For easier acces I have in my init.lua:
of course you could do the same with autocmd
``` lua
vim.api.nvim_create_user_command("UpdateProjectList", "!python /path/to/python_script.py", {})
vim.api.nvim_set_keymap('n', "<leader>p", '<cmd>e /path/to/your/vault/Projektlist_dynamic.md<cr>', {silent = true})
vim.api.nvim_set_keymap('n', "<leader>U", '<cmd>UpdateProjectList<cr>', {silent = true})
```
### 3. Further ideas:

1. If you want to do more complicated queries, you could load the frontmatter dictionary into a pandas dataFrame
(Look into pandas.DataFrame.from_dict)
2. The above only looks into the yaml-frontmatter, But Dataview also looks for key:: value pairs in the body.
Obviously `#tags` are not taken into consideration, but Dataview looks for them, too.
3. Bonus: Python's PTable can also sort data.

**Comment by jkmacc** [comment]
I like that this approach is understandable and easy for most folks to extend.  The closest I've found to dataview in Neovim is [dynomark](https://github.com/k-lar/dynomark) and it's [plugin](https://github.com/k-lar/dynomark.nvim), but it's written in Go and I'm not familiar with that language.  I agree that a dataview alternative is sorely needed!

**Reply by psxvoid** [reply]
> I agree that a dataview alternative is sorely needed!

Or a runtime emulation layer and plugin system that can install official plugins.

**Reply by Dardanos-Aeolus** [reply]
Oh, nice.
I didn't know about dynomark, but it looks awesome!
It seems pretty powerful.
However, from what I see on the project page, it looks that Windows is not supported, yet.

**Comment by Dardanos-Aeolus** [comment]
Update:
Now, there is krafna:
[Krafna](https://github.com/7sedam7/krafna)

It's a command-line tool, it can be integrated into Neovim, with the perec Plugin from the same author:
[perec.nvim](https://github.com/7sedam7/perec.nvim)

Though I'm not quite sure if Perec works well in parallel to obsidian.nvim: Some of the feature of perec are redundant to obsidian.nvim

**Reply by justyn** [reply]
I can't tell from the Krafna page - is it compatible with the dataview syntax, so that an obsidian note with embedded dataview syntax can be run in neovim?

**Comment by troutpies** [comment]
just brainstorming here...

I currently use markdown-toc to update a TOC for my current file on save. It updates the text between two specific HTML comments (I have it set to `<!-- mtoc-start -->` and `<!-- mtoc-end -->`).

Perhaps a dataview alternative could take the form of a supplemental plugin, which queries a list of files (interacting with obsidian.nvim), and then inserts a list of links. Arguments to the query could be specified in the comment line, or the user could define some presets in their configuration. It might look something like

```markdown
<!-- mview-start tags: [project/open, !archive] -->
<!-- mview-end -->
```
or
```markdown
<!-- mview-start name: my-special-query -->
<!-- mview-end -->
```

Queries might be defined by callback functions, either using some kind of API, or user-written lua.
